### PR TITLE
Add code for Twitch login

### PR DESCRIPTION
**Pull request by lipis**
Test it using: `http://127.0.0.1:8080/` and not the `http://localhost:3000`

### CODE REVIEW

File: main/auth/twitch.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+
+@app.route('/api/auth/callback/twitch/')
+def twitch_authorized():
+  id_token = twitch.authorize_access_token()
+  if id_token is None:
+    flask.flash('You denied the request to sign in.')
+    return flask.redirect(util.get_next_url())
+
+  me = twitch.get('people/~:(id,email-address)')  # Need to change this line
```
Comment (rbubley) [inline main/auth/twitch.py:37]: Maybe ` me = twitch.get('user')`

File: main/auth/twitch.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def retrieve_user_from_twitch(response):
+  auth_id = 'twitch_%s' % response['id']
+  user_db = model.User.get_by('auth_ids', auth_id)
+  if user_db:
+    return user_db
+
+  name = response['display_name']
+  username = response.get('preferred_username', '')
```
Comment (rbubley) [inline main/auth/twitch.py:54]: Maybe:` username = response.get('display_name', '')`

File: main/auth/twitch.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+twitch_config = dict(
+  access_token_method='POST',
+  access_token_url='https://id.twitch.tv/oauth2/token',
+  api_base_url='https://api.twitch.tv/helix/',
+  authorize_url='https://id.twitch.tv/oauth2/authorize',
+  client_id=config.CONFIG_DB.twitch_client_id,
+  client_secret=config.CONFIG_DB.twitch_client_secret,
+  request_token_params={
+    'scope': 'user:read:email;user:read:stream_key',
```
Comment (lipis) [inline main/auth/twitch.py:22]: @rbubley would that be the correct way to ask for email and the handle? 

https://dev.twitch.tv/docs/authentication#scopes

**Comment by rbubley** [comment]
@lepture -- any insight? When we call `authorize_access_token()` we get a 400 error, stating `no client id`.

File: main/auth/twitch.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+
+def retrieve_user_from_twitch(response):
+  auth_id = 'twitch_%s' % response['id']
+  user_db = model.User.get_by('auth_ids', auth_id)
+  if user_db:
+    return user_db
+
+  name = response['display_name']
+  username = response.get('display_name', '')
```
Comment (lipis) [inline main/auth/twitch.py:63]: we should try to get the Twitch handle here.

File: main/auth/twitch.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  },
+)
+
+twitch = auth.create_oauth_app(twitch_config, 'twitch')
+
+def twitch_compliance_fix(url, headers, data):
+  headers["Client-ID"] = config.CONFIG_DB.twitch_client_id
+  return url, headers, data
+
+twitch.pre_request = twitch_compliance_fix
```
Comment (lepture) [inline main/auth/twitch.py:34]: what is this `pre_request`? In Authlib, we need to register it as `compliance_fix` like https://github.com/authlib/loginpass/blob/master/loginpass/twitch.py#L54

File: main/auth/twitch.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+from main import app
+
+twitch_config = dict(
+  access_token_method='POST',
+  access_token_url='https://id.twitch.tv/oauth2/token',
+  api_base_url='https://api.twitch.tv/helix/',
+  authorize_url='https://id.twitch.tv/oauth2/authorize',
+  client_id=config.CONFIG_DB.twitch_client_id,
+  client_secret=config.CONFIG_DB.twitch_client_secret,
+  request_token_params={
```
Comment (lepture) [inline main/auth/twitch.py:21]: I think this `request_token_params` should be `client_kwargs`.

Comment (lipis) [reply main/auth/twitch.py:21]: That worked.. and now we have a new error!

File: main/auth/twitch.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  return auth.signin_user_db(user_db)
+
+
+@app.route('/signin/twitch/')
+def signin_twitch():
+  return auth.signin_oauth(twitch)
+
+
+def retrieve_user_from_twitch(response):
+  auth_id = 'twitch_%s' % response['id']
```
Comment (lipis) [inline main/auth/twitch.py:57]: The error appears here and when I'm printing the `response` I'm getting:

```
{u'status': 404, u'error': u'Not Found', u'message': u''}
```

/cc @lepture @rbubley

File: main/auth/twitch.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+twitch.pre_request = twitch_compliance_fix
+
+@app.route('/api/auth/callback/twitch/')
+def twitch_authorized():
+  id_token = twitch.authorize_access_token()
+  if id_token is None:
+    flask.flash('You denied the request to sign in.')
+    return flask.redirect(util.get_next_url())
+
+  me = twitch.get('user')
```
Comment (lipis) [inline main/auth/twitch.py:43]: Actually the problem is here.. there is no `user` in `twitch`. How can we list what's there?

File: main/auth/twitch.py (hunk 5 above)
Comment (lipis) [reply main/auth/twitch.py]: I removed it completely @rbubley.. I think it wasn't called at all.

File: main/auth/twitch.py (hunk 8 above)
Comment (lepture) [reply main/auth/twitch.py:40]: I think the endpoint is “users”. You can learn it from loginpass

File: main/auth/twitch.py (hunk 5 above)
Comment (lepture) [reply main/auth/twitch.py]: But you still need to add a complaint fix. You can learn it from loginpass.

**Comment by lipis** [comment]
I can see https://github.com/gae-init/gae-init/compare/twitch?expand=1 more commits there than here!

**Comment by lipis** [comment]
Something was wrong with this PR... made a new one: #1496

**Comment by rn4n** [comment]
@rbubley I'm getting the same error, were you able to authenticate?

```
token = await oauth.twitch.authorize_access_token(request)

ipdb> token
{'status': 400, 'message': 'missing client id'}
```

**Comment by rbubley** [comment]
I thought @lipis got it working in #1496.

**Pull request closed without merging**